cedar-policy-core/src/entities.rs

```diff
@@ -48,6 +52,9 @@ pub struct Entities {
     #[serde_as(as = "Vec<(_, _)>")]
     entities: HashMap<EntityUID, Entity>,
 
+    #[serde(skip)]
+    evaluated_entities: Option<EvaluatedEntities>,
```

Comment:
Oh no my merge conflicts :(

We are solving essentially the same problem in pretty different ways. I've been trying to keep my fork up to date with cedar `main` so that merging is/will be less troublesome, but I think that this will force a split because it solves an existing problem in a different way from how I solved it.

I made `Entity` and `Entities` generic around a type `T` which is the type of attributes. Then, `T` can take the value of either `RestrictedExpr` or `PartialValue`  (with `RestrictedExpr` the default to prevent breaking changes). I then changed the evaluator to only work with `EntityDatabase` (which is implemented only for `Entities<PartialValue>`, not `Entities<RestrictedExpr>`, because the `get` method needs to return attributes of type `PartialValue`). Finally I modified `is_authorized` to convert an `Entities<RestrictedExpr>` to an `Entities<PartialValue>` and duplicated the API for `is_authorized` to take in `Entities<PartialValue>`  instead of`Entities<RestrictedExpr>` in case the user wants to do this parsing/caching themselves (the goal is eventually the duplicated API will be the default and the other/old API will be deprecated/removed).

Could you comment on the pros/cons of this approach? Besides minor performance benefits because you don't duplicate the `HashMap`, the biggest reason I need this is there is an explicit type for entities with evaluated attributes, which can implement `EntityDatabase`. The major drawback is it duplicates all existing API to take in the new type instead of just adding a method to `Entities` like yours does.

Reply:
Discussed in other comments (eg [this one](https://github.com/cedar-policy/cedar/pull/230#discussion_r1276666878)) -- I'd also love for "entities with evaluated attributes" to be an explicitly different type from "entities without evaluated attributes", but Aaron's solution in this PR avoids changing the `is_authorized` signature for back-compatibility.  Later, on next major, we could maybe make the type-level changes and change the signature of `is_authorized`.